routing/index_graph_starter.cpp

```diff
@@ -1,192 +1,423 @@
 #include "routing/index_graph_starter.hpp"
 
+#include "routing/fake_edges_container.hpp"
+
 #include "geometry/distance.hpp"
 
+#include "base/math.hpp"
+
 namespace
 {
 using namespace routing;
 using namespace std;
 
-Junction CalcProjectionToSegment(Segment const & segment, Junction const & junction,
+Segment GetReverseSegment(Segment const & segment)
+{
+  return Segment(segment.GetMwmId(), segment.GetFeatureId(), segment.GetSegmentIdx(),
+                 !segment.IsForward());
+}
+
+Junction CalcProjectionToSegment(Segment const & segment, m2::PointD const & point,
                                  WorldGraph & graph)
 {
+  auto const & begin = graph.GetJunction(segment, false /* front */);
+  auto const & end = graph.GetJunction(segment, true /* front */);
   m2::ProjectionToSection<m2::PointD> projection;
-  projection.SetBounds(graph.GetPoint(segment, false /* front */),
-                       graph.GetPoint(segment, true /* front */));
-  return Junction(projection(junction.GetPoint()), junction.GetAltitude());
+  projection.SetBounds(begin.GetPoint(), end.GetPoint());
+  auto projectedPoint = projection(point);
+  auto const distBeginToEnd = graph.GetEstimator().CalcLeapWeight(begin.GetPoint(), end.GetPoint());
+
+  double constexpr kEpsMeters = 2.0;
+  if (my::AlmostEqualAbs(distBeginToEnd, 0.0, kEpsMeters))
+    return Junction(projectedPoint, begin.GetAltitude());
+
+  auto const distBeginToProjection =
+      graph.GetEstimator().CalcLeapWeight(begin.GetPoint(), projectedPoint);
+  auto const altitude = begin.GetAltitude() + (end.GetAltitude() - begin.GetAltitude()) *
+                                                  distBeginToProjection / distBeginToEnd;
+  return Junction(projectedPoint, altitude);
+}
+
+template<typename K, typename V>
+bool IsIntersectionEmpty(map<K, V> lhs, map<K, V> rhs)
```

Comment:
It would be nice to pass maps by const ref.

Reply:
ups, done

Reply:
You may do this more efficiently (O(n + m)) and in a more concise way:
```c++
size_t CountIntersection(map<K, V> const & lhs, map<K, V> const & rhs)
{
  auto const counter = set_intersection(lhs.begin(), lhs.end(), rhs.begin(), rhs.end(),
                                        CounterIterator(), my::LessBy(&pair<K, V>::first));
  return counter.GetCount() == 0;
}
```

Reply:
done